Add deploy_code_to_rpi.sh script

KaneBetter opened the pull request:
Fix #18 
Might Fix #20 
 I suggest moving videos into `/home/pi/Videos/` for autostart use.
The deploy_code_to_rpi.sh can extract the host IP from max-init.txt file and then use pdsh command ssh into each host and then run the command mentioned in README.md to deploy code.

KaneBetter committed:
AddedScriptForContinuousDeployment

SSSSam141 committed:
Update readme.md with OS installation. (#17)

KaneBetter committed:
Replace VLC with OMXPlayer (#19)

* Used Omxplayer in system 10
* Add shell script for build omxplayer
* Add set rate feature
* Added logging format and TODO list
* Remove vlc and change README.md

omarcostahamido committed:
Lib folder (#23)

- this fixes #21 
- users will only have to open `pvm.maxproj` now

* move max files into `lib` folder

* Create pvm.maxproj

* Update README.md

- added reference to `pvm.maxproj`
- moved long explanation of the max control interface to the structure section.

omarcostahamido committed:
Creative examples (#24)

- this fixes #22 

* Create examples.maxproj

* Create #01.maxpat

- example with: sync video in 2 devices

* Create #02.maxpat

- example with: sync video in 4 devices

* Create #03.maxpat

- example with: offset video in 2 devices

* Create #04.maxpat

- example with: offset video in 4 devices

* Create #05.maxpat

- example with: different speeds in 4 devices

* create examples index patch

* add `Examples` section

- in order to mention the examples folder and how to use it.

KaneBetter committed:
Revise pdsh script

KaneBetter committed:
Update README.md and deploy code script

I suggest to move videos into `/home/pi/Videos/` for autostart use.
The deploy_code_to_rpi.sh can extract host ip from max-init.txt file and then
use pdsh command ssh into each host and then run the command mentioned in README.md to deploy code.

KaneBetter committed:
Merge branch 'main' into AddCDscript

omarcostahamido commented on README.md:
```diff
@@ -132,6 +138,23 @@ _           | `host.py`                   | (to be removed) control a remote dev
+Use the [deploy_code_to_rpi.sh](https://github.com/omarcostahamido/PVM/blob/AddCDscript/deploy_code_to_rpi.sh) script.
+But there are two prerequisites:
+
+One is the need to set up passwordless ssh access.
```
this sounds a little bit like trouble on the long road. Imagining these devices to be installed on public locations with public/shared networks, might expose this considerably. We don't want to run into a situation where a multiscreen installation gets hacked. Are you sure it is not possible to work with passwords?

omarcostahamido replied:
(so yes, disregard this first review comment)

omarcostahamido commented on README.md:
```diff
@@ -132,6 +138,23 @@ _           | `host.py`                   | (to be removed) control a remote dev
+But there are two prerequisites:
+
+One is the need to set up passwordless ssh access.
+For the passwordless ssh access please follow [here](https://danidudas.medium.com/how-to-connect-to-raspberry-pi-via-ssh-without-password-using-ssh-keys-3abd782688a).
```
ah ok, it is using private-public rsa keys. That's much better! Thanks.
maybe you should change the text slightly to clarify that the goal is not simply to remove the password, it is also to impose a rsa public-private key.
I've saved this article (nice one) on [wayback machine](https://web.archive.org/web/20221029175737/https://danidudas.medium.com/how-to-connect-to-raspberry-pi-via-ssh-without-password-using-ssh-keys-3abd782688a).

KaneBetter replied:
Now I hope this script will alleviate our deployment problems during development testing. This way the secret key form will reduce the time we spend on testing.
If there are future deployment issues during production, perhaps a cycle update approach could be considered.

omarcostahamido commented on README.md:
```diff
@@ -132,6 +138,23 @@ _           | `host.py`                   | (to be removed) control a remote dev
+One is the need to set up passwordless ssh access.
+For the passwordless ssh access please follow [here](https://danidudas.medium.com/how-to-connect-to-raspberry-pi-via-ssh-without-password-using-ssh-keys-3abd782688a).
+
+Second is the need to set the autostart.
```
as described above, right? can you please hyperlink it?
Also please clarify that this means setting up autostart on the rpi devices, not the control machine.

omarcostahamido replied:
(referring to line 148)

KaneBetter replied:
Added hyperlink.

omarcostahamido replied:
@KaneBetter where?
(maybe some commits that are not showing up here on github yet?)

omarcostahamido commented on README.md:
```diff
@@ -132,6 +138,23 @@ _           | `host.py`                   | (to be removed) control a remote dev
+
+`cd PVM && git pull`
+
+`sudo reboot`
```
restarting the control machine?

KaneBetter replied:
I have thought about this question for a long time. I realized that our Raspberry Pi must eventually be set to autostart, so I wanted our test environment to be in an autostart environment as well, so after multiple options, I chose this reboot format.

omarcostahamido replied:
as per line 148 `autostart` is a requirement on the rpi devices for this `deploy_code_to_rpi.sh` script to work as intended. What I am not clear about is why are you restarting the control machine at this point.

KaneBetter replied:
There are two ways we want to use the new code, one is to close the existing pvm program and restart it. The second one is to restart the rpi. The reason why we choose the second one here is that the second one is more in line with our actual scenario. 
And I found in the actual test, the Raspberry Pi reboot time is short, less than 1 minute, acceptable.

omarcostahamido replied:
@KaneBetter got it. As per my comment [here](https://github.com/omarcostahamido/PVM/pull/26#discussion_r1008914713) I had not looked well at `deploy_code_to_rpi.sh` and I thought these two lines were being executed on the control machine. Thank you.

omarcostahamido commented on README.md:
```diff
@@ -132,6 +138,23 @@ _           | `host.py`                   | (to be removed) control a remote dev
+
+Second is the need to set the autostart.
+
+Once you're done, run the command on your computer.
```
would be nice to also clarify here that it will deploy the latest (maybe unstable?) version of the software on all the devices whose IP number appears on the .txt file.

KaneBetter replied:
Added comment.

omarcostahamido commented on host_ip.txt:
```diff
@@ -0,0 +1,6 @@
+192.168.1.108
+192.168.0.102
+192.168.0.103
+192.168.0.104
+192.168.0.100
+192.168.0.106
```
it does not make sense to push this file to github since it will always be programmatically recreated.
I would suggest adding `host_ip.txt` to the gitignore.
Please also ensure that `deploy_code_to_rpi.sh` also deletes this file before it tries to recreate it!

KaneBetter replied:
Added host_ip.txt in .gitignore.

omarcostahamido commented on pvm.py:
```diff
@@ -21,6 +21,8 @@ def _init_logger():
 # _logger.info('App started in %s', os.getcwd())
 # _logger.debug('App started in %s', os.getcwd())
 
+# Place your videos in this folder for autostart
+PEFIX_PATH = "/home/pi/Videos/"
```
missing instructions about this on the readme!

KaneBetter replied:
Added Videos section in README

omarcostahamido commented on README.md:
```diff
@@ -10,6 +10,12 @@ Pi Video Machine - a scalable, synchronized, and networked-controlled, raspberry
 - TODO: Create another python file to control two display
 - TODO: create a isFileSet flag
 
+- TODO: rename the variable
+- TODO: rewrite logging
+- TODO: rewrite logic between commands
+- TODO: Create another python file to control two display
+- TODO: create a isFileSet flag
```
is it me or is there a repetition here?

KaneBetter replied:
Deleted

omarcostahamido reviewed (changes requested):
please review comments below. Thank you!

KaneBetter committed:
Revised README.md

- Remove VLC in README.md and add OMXPlayer
- Remove duplication in README.md
- Change some instrucation in installation

omarcostahamido commented on README.md:
```diff
@@ -23,6 +17,13 @@ Pi Video Machine - a scalable, synchronized, and networked-controlled, raspberry
 ## Installation
 ### Prerequisites
+
+### OMXPlayer
+
+To build the customized OMXPlayer, run the command below:
+
+`./build_omxplayer.sh`
```
Thank you for adding this. It is definitely a needed subsection in `## Installation`.
However, it is expected that users will be building OMXPlayer **after** they flashed the Raspberry Pi OS and they are able to access the rpi device.

KaneBetter replied:
Moved this section to after the OS installation

omarcostahamido commented on README.md:
```diff
@@ -54,44 +55,45 @@ To install this to your Raspberry Pi, you need the following things:
+# Create a virtual environment
+python3 -m venv PVM
 
-`pip install --upgrade pip setuptools`
```
simply delete this?
(Actually, I wonder if there is a way to simply upgrade these from within the `requirements.txt`)
In my experience I've noted that not updating pip and setuptools on a venv sometimes leads to problems when installing packages...

KaneBetter replied:
Readded

omarcostahamido replied:
🙌

omarcostahamido commented on README.md:
```diff
@@ -145,15 +146,18 @@ But there are two prerequisites:
 
-`cd PVM && git pull`
+# Download the latest code on RPIs with the IP address in host_ip.txt.
+cd PVM && git pull
 
-`sudo reboot`
+sudo reboot
```
so you're saying that these 2 lines of code will be executed on the rpi devices even though they are typed on the control machine?
Ok, got it, the script just opens the connection, it doesn't actually do the heavy work of actually updating the devices! It feel that these lines of code could very well be already part of the `deploy_code_to_rpi.sh` script, no?

KaneBetter replied:
Yes, you are right about the idea. The control PC can reboot all RPIs when typing `reboot` in its console.
I try to do it in the script. Unfortunately, pdsh will create a new interactive session that will not receive the command from the script. 
But I will still work on this if a better method comes up.

omarcostahamido replied:
Thank you @KaneBetter ! 😄

omarcostahamido commented on README.md:
thx for the cleanup from line 166 until here!

omarcostahamido reviewed (approved):
Thank you for addressing the comments and doing some cleanup.
I've added some further comments that you can take a look at below.
In any case LGTM, and you shall be able to merge when you believe it is best.
Thank you.

KaneBetter committed:
Move OMXPlayer section
